core/src/main/java/org/apache/calcite/rel/metadata/RelMdMinRowCount.java

```diff
@@ -61,7 +61,12 @@ public Double getMinRowCount(Union rel, RelMetadataQuery mq) {
         rowCount += partialRowCount;
       }
     }
-    return rowCount;
+
+    if (rel.all) {
+      return rowCount;
+    } else {
+      return Math.min(rowCount, 1d);
+    }
```

Comment:
Maybe there is a potential improvement here: if rel.all == false, and once we find the current value of rowCount greater than 1d, we can stop the loop early and directly return 1d.

It also makes the code less readable, so I am not sure if it is worth it.

Reply:
Considering there would not be a huge overhead, I prefer it more readable.

Reply:
Sounds reasonable.